Project sorting

AlessioGr opened the pull request:
![image](https://user-images.githubusercontent.com/70709113/152452550-b74fb9ec-ea0a-4e9c-a39e-116faab1dad4.png)
This adds the sorting system which seems to work just fine in my testing.

However, for that to work, I added the sorting stuff to the RequestPagination. Using the ProjectSortingStrategy for that would probably be better, but I'm not sure how I can change that.

When I change the sorting in the frontend, the RequestPaginationResolver.resolveArgument method is called, where I can snack the sort type from the NativeWebRequest webRequest.

However, it's not apparent to me where the ProjectSortingStrategy is created / where it communicated with the frontend there.

Machine-Maker commented on frontend/locales/en.ts:
```diff
@@ -91,6 +91,13 @@ export default {
       auth: 'Authentication Portal'
     }
   },
+  filters: {
```
these aren't filters

AlessioGr replied:
I renamed the key to "sorting"

Machine-Maker commented on src/main/java/io/papermc/hangar/service/api/ProjectsApiService.java:
```diff
@@ -81,8 +98,45 @@ public Project getProject(String author, String slug) {
             ordering = orderingFirstHalf + relevance;
         }
 
+        if(paginationSortValue != null && !paginationSortValue.isBlank()){
+            String relevance = "ASC";
+            String orderingFirstHalf;
+
+            switch (sort) {
+                case STARS -> {
+                    orderingFirstHalf = "hp.stars ";
+                    relevance = "DESC";
+                }
+                case DOWNLOADS -> {
+                    orderingFirstHalf = "hp.downloads ";
+                    relevance = "DESC";
+                }
+                case VIEWS -> {
+                    orderingFirstHalf = "hp.views ";
+                    relevance = "DESC";
+                }
+                //case NEWEST: orderingFirstHalf ="((EXTRACT(EPOCH FROM hp.created_at) - 1609459200) / 86400) *"; break;
+                case NEWEST -> {
+                    orderingFirstHalf = "newest_double ";
+                    relevance = "DESC";
+                }
+                case UPDATED -> {
+                    orderingFirstHalf = "last_updated_double ";
+                    relevance = "DESC";
+                }
+                case ONLY_RELEVANCE -> orderingFirstHalf = "";
+                case RECENT_DOWNLOADS -> orderingFirstHalf = "hp.recent_views ";
+                case RECENT_VIEWS -> orderingFirstHalf = "hp.recent_downloads ";
+                default -> orderingFirstHalf = " "; // Just in case and so that the ide doesn't complain
+            }
+            ordering = orderingFirstHalf + relevance;
+
+        }
+
```
This should be implemented in a similar way to fitlers, with a registry and using annotation to define valid sorters for api requests. This abstracts sorting to be valid for more things than just this one api request.

Machine-Maker reviewed (changes requested):
What did you do to the `yarn.lock` file?

I don't think this is the correct way to go about implementing sorters. The ProjectSortingStrategy enum should be removed tho. Its a relic from Ore. But they way you have implemented this only allows for sorting by one thing at a time, not multiple.

AlessioGr commented:
> What did you do to the `yarn.lock` file?

I fixed the yarn.lock and excluded it from my commits. For some reason, my IDE is not happy with the default yarn.lock

> I don't think this is the correct way to go about implementing sorters. The ProjectSortingStrategy enum should be removed tho. Its a relic from Ore. But they way you have implemented this only allows for sorting by one thing at a time, not multiple.

Regarding "But they way you have implemented this only allows for sorting by one thing at a time, not multiple" how would sorting by multiple things work? From how I'd understand it, you can sort by latest update and stars, but not both.

For the `ProjectSortingStrategy`, should it be removed from the `getProjects` method completely?

What I currently don't understand is this method from the `ProjectsController.java`:
```java
 @Override
    @ApplicableFilters({ProjectCategoryFilter.class, ProjectPlatformFilter.class, ProjectAuthorFilter.class, ProjectQueryFilter.class, ProjectTagFilter.class})
    public ResponseEntity<PaginatedResult<Project>> getProjects(String q, ProjectSortingStrategy sort, boolean orderWithRelevance, @NotNull RequestPagination pagination) {
        return ResponseEntity.ok(projectsApiService.getProjects(q, sort, orderWithRelevance, pagination));
    }
```
Could you explain to me how it works? My IDE cannot find any usages for this method, and I don't know where / how it's called. How do these annotations here work, and where does its ProjectSortingStrategy come from?

Machine-Maker commented on frontend/types/enums.ts:
```diff
@@ -5,6 +5,14 @@ export enum RoleCategory {
     ORGANIZATION = 'organization',
 }
 
+export enum ProjectSort {
+    MOST_STARS = 'most_stars',
+    MOST_DOWNLOADS = 'most_downloads',
+    MOST_VIEWS = 'most_views',
+    NEWEST = 'newest',
+    RECENTLY_UPDATED = 'recently_updated',
+}
+
```
I would just have this as a local enum inside the component file. All the rest of the enums in this file have a mirrored enum in the java project, but this one doesn't. And like I said earlier, the ProjectSortingStrategy should be removed at some point. So I can't see this being used anywhere outside of that component.

AlessioGr commented:
Alright, I think I implemented this the way you wanted – using the `SorterRegistry`. From my testing, it works just fine!

I have also removed the `ProjectSortingStrategy`.
